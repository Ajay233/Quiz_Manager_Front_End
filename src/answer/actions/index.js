import { getUsingParams, post, put, del } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { hideModal } from '../../modal/actions'
import { sessionExpired } from '../../utils/session'
import history from '../../history'

// create
export const addAnswer = (data, jwt) => {
  return (dispatch) => {
    return post("answer/create", [data], jwt).then((response) => {
      dispatch({
        type: "ADD_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer created", "success", true));
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        console.log(error.response);
        dispatch(setNotification("Error - unable to create answer", "error", true));
      }
    });
  }
}

// read
export const getAnswers = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("answer/findByQuestionId", param, jwt).then((response) => {
      dispatch({
        type: "SET_ANSWERS",
        payload: response.data
      });
    }).catch((error) => {
      dispatch(setNotification("Error retrieving answers", "error", true));
    });
  }
}

// update (there was a call to setCurrentAnswer but have removed this as it's not needed)
export const updateAnswer = (body, jwt) => {
  return (dispatch) => {
    return put("answer/update", [body], jwt).then((response) => {
      dispatch({
        type: "UPDATE_ANSWER",
        payload: response.data[0]
      })
      history.push("/editQuestion");
      dispatch(setNotification("Answer updated", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - unable to update answer", "error", true));
      }
    });
  }
}

// delete
export const deleteAnswer = (config, jwt) => {
  return (dispatch) => {
    return del("answer/delete", config, jwt).then((response) => {
      dispatch(hideModal())
      dispatch({
        type: "DELETE_ANSWER",
        payload: config.data[0]
      })
      dispatch(setNotification("Answer deleted", "success", true));
    }).catch((error) => {
      console.log(error.response);
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(hideModal())
        dispatch(setNotification("Error - Unable to delete this answer", "error", true))
      }
    });
  }
}

export const setCurrentAnswer = (answer) => {
  return {
    type: "SET_CURRENT_ANSWER",
    payload: answer
  }
}
