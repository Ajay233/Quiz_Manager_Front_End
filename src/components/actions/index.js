export const setToTopButton = (show=false) => {
  return {
    type: "SHOW_TO_TOP_BUTTON",
    payload: show
  }
}

export const setUrl = (url="") => {
  return {
    type: "SET_URL",
    payload: url
  }
}

export const enableLightMode = () => {
  return {
    type: "SET_THEME",
    payload: false
  }
}

export const enableDarkMode = () => {
  return {
    type: "SET_THEME",
    payload: true
  }
}

export const setLoaderState = (show=false, message="", label="") => {
  return {
    type: "SET_LOADER_STATE",
    payload: {
      show: show,
      message: message,
      label: label
    }
  }
}
