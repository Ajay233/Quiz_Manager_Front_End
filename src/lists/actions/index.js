import { get } from '../../axiosRequests/requests'
import { sessionExpired } from '../../utils/session'

export const getCategories = (endpoint, jwt) => {
  return (dispatch) => {
    get(endpoint, jwt).then((response) => {
      dispatch({
        type: "SET_CATEGORIES",
        payload: response.data
      })
    }).catch((error) => {
      console.log(error.response)
      if(error.response.status === 403){
        sessionExpired(dispatch);
      }
    })
  }
}
