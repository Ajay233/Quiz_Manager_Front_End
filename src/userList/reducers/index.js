import { listOfUsers } from '../factory/userListFactory'

export const setUserListReducer = (userList=listOfUsers, action) => {
  switch (action.type) {
    case "SET_USER_LIST": return action.payload;
    case "DELETE_USER_FROM_LIST": return userList.filter(user => user.id !== action.payload);
    case "CLEAR_USER_LIST": return action.payload;
    default: return userList;
  }
}
