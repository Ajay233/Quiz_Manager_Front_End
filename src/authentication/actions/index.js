import { post } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import history from '../../history'

export const setUser = (endpoint, loginDetails) => {

  return (dispatch) => {

    post(endpoint, loginDetails).then((response) => {
      dispatch({
        type: "SET_USER_LOGGED_IN",
        payload: {
          id: response.data.user.id,
          forename: response.data.user.forename,
          surname: response.data.user.surname,
          email: response.data.user.email,
          permission: response.data.user.permission,
          verified: response.data.user.verified,
          jwt: response.data.jwt,
          loggedIn: true,
          response: response
        }
      })
      history.push("/")
      dispatch(setNotification(`Welcome back ${response.data.user.forename}`, "loginSuccess", true))
    }).catch((error) => {
      if(error.response.data === "NOT VERIFIED"){
        const verifyMsg = "Your email has not yet been veirified.  You will need to verify your email before you can log in"
        dispatch(setNotification(verifyMsg, "warning", true))
      } else {
        dispatch({
          type: "SET_USER_LOGGED_IN",
          payload: {
            error: error.response
          }
        });
        dispatch(setNotification("The username or password you entered was incorrect", "error", true));
      }
    })
  }
}

export const logOutUser = () => {
  return {
    type: "LOG_OUT_USER",
    payload: {
      forename: "",
      surname: "",
      email: "",
      permission: "",
      verified: "",
      jwt: "",
      loggedIn: false
    }
  }
}

export const setVerficationProcess = (endpoint, verificationDetails) => {

  return (dispatch) => {

    post(endpoint, verificationDetails).then((response) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "completed",
          token: "",
          error: {}
        }
      })
    }).catch((error) => {
      dispatch({
        type: "SET_VERIFY_PROCESS_STATUS",
        payload: {
          completionStatus: "not verified",
          token: verificationDetails.token,
          error: error.response
        }
      })
    })
  }
}

export const setRedirect = (status) => {
  return {
    type: "SET_REDIRECT",
    payload: {
      status: status
    }
  }
}

export const logOut = () => {
  return {
    type: "RESET_APP"
  }
}

// export const getUsers = (endpoint, header) => {
//   return {
//     type: "GET_USER",
//     payload: {
//       userId
//   }
// }
