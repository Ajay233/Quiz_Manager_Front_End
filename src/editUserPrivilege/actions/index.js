import { getUsingParams } from '../../axiosRequests/requests'
import { setNotification } from '../../notifications/actions'
import { sessionExpired } from '../../utils/session'

export const setUserResults = (param, jwt) => {
  return (dispatch) => {
    return getUsingParams("users/findByEmail", param, jwt).then((response) => {
      dispatch({
        type: "SET_USER_RESULTS",
        payload: response.data
      })
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        const msg = error.response.data;
        dispatch(setNotification(msg, "error", true));
        console.log(error.response)
      }
    });
  }
}

export const clearUserResults = () => {
  return {
    type: "CLEAR_USER_RESULTS",
    payload: {}
  }
}
