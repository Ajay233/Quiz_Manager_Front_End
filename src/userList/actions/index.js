import { get, del } from '../../axiosRequests/requests'
import { sessionExpired } from '../../utils/session'
import { setNotification } from '../../notifications/actions'

export const setUserList = (endpoint, data, token) => {
  return (dispatch) => {
    return get(endpoint, data, token).then((response) => {
      dispatch({
        type: "SET_USER_LIST",
        payload: response.data
      })
    }).catch((error) => {
      if(error.response.status === 403){
        sessionExpired(dispatch);
      } else {
        dispatch(setNotification("Error - unable to get user list", "error", true));
      }
    })
  }
}

export const deleteUser = (config, jwt) => {
  return (dispatch) => {
    del("users/deleteAccount", config, jwt).then((response) => {
      dispatch({
        type: "DELETE_USER_FROM_LIST",
        payload: config.data.id
      })
      dispatch(setNotification("Account deleted", "success", true));
    }).catch((error) => {
      console.log(error.response)
      dispatch(setNotification("Error - Account could not be deleted", "error", true));
    });
  }
}

export const clearUserList = () => {
  return {
    type: "CLEAR_USER_LIST",
    payload: []
  }
}
