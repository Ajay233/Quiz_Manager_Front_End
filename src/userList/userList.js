import React from 'react';
import { connect } from 'react-redux'
import Notification from '../notifications/notifications'
import { setUserList, clearUserList, deleteUser } from './actions'

import '../stylesheets/userList.css'

export class UserList extends React.Component {

  componentDidMount(){
    this.props.setUserList("users", this.props.userData.jwt);
    document.documentElement.scrollTop = 0;
  }

  componentWillUnmount(){
    this.props.clearUserList()
  }

  deleteUser = (id) => {
    const { userData, deleteUser } = this.props
    const config = { data: { id: id } }
    deleteUser(config, userData.jwt)
  }

  renderList = () => {
    const { listOfUsers } = this.props
    let userList = listOfUsers.map(user => {
      return(
        <div key={listOfUsers.indexOf(user)} className="userDetails userContainer">
          <div className="forename">{user.forename}</div>
          <div className="surname">{user.surname}</div>
          <div className="email">{user.email}</div>
          <div className="permission">{user.permission}</div>
          <div className="verified">{ this.renderVerified(user.verified) }</div>
          <div className="deleteUser" onClick={() => {this.deleteUser(user.id)}}><i className="fas fa-trash-alt red"></i> Delete</div>
        </div>
      );
    })
    return userList
  }

  renderVerified = (verified) => {
    return verified ? <i className="far fa-check-circle green"></i> : <i className="far fa-times-circle red"></i>
  }

  renderHeaders = () => {
    return(
      <div className="userDetailHeaders userHeadings">
        <div className="forenameHeader">First Name</div>
        <div className="surnameHeader">Surname</div>
        <div className="emailHeader">Email</div>
        <div className="permissionHeader">Permission</div>
        <div className="verifiedHeader">Verified?</div>
        <div className="deleteUserHeader"></div>
      </div>
    );
  }

  render(){
    return(
      <div className="componentContainer">
        <Notification />
        <div className="title-large">List of all users</div>
        {this.renderHeaders()}
        {this.renderList()}
      </div>
    );
  }
}

export const mapStateToProps = (state) => {
  return {
    userData: state.userData,
    listOfUsers: state.listOfUsers
  };
}

export default connect(mapStateToProps, { setUserList, clearUserList, deleteUser })(UserList)
