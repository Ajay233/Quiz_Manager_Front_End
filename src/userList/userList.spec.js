import React from 'react'
import ReactDOM from 'react-dom'
import { UserList, mapStateToProps } from './userList'
import { Router } from 'react-router-dom'
import { Provider } from 'react-redux'
import configureStore from 'redux-mock-store'
import mockAxios from 'jest-mock-axios'
import { render, fireEvent, cleanup } from '@testing-library/react'
import renderer from 'react-test-renderer'
import history from '../history'

import { setUserList, clearUserList } from './actions'
import { setNotification } from '../notifications/actions'

jest.mock('./actions')
jest.mock('../notifications/actions')

const mockStore = configureStore({})

describe("mapStateToProps", () => {
  it("should map state to props", () => {
    const appState = {
      userData: {id: 1, jwt: "jwt"},
      listOfUsers: [
        {id: 1},
        {id: 2}
      ]
    }

    const componentState = mapStateToProps(appState)
    expect(componentState).toEqual(appState)
  })
})

describe("UserList", () => {

  let store;
  let listOfUsers;
  let userData;
  let notificationData;

  beforeEach(() => {
    store = mockStore({
      notificationData: {
        message: "",
        type: "",
        show: false,
        timed: true
      }
    })

    listOfUsers = [
      {
        forename: "testForename1",
        surname: "testSurname1",
        email: "testEmail1",
        permission: "USER",
        verified: "true"
      },
      {
        forename: "testForename2",
        surname: "testSurname2",
        email: "testEmail2",
        permission: "USER",
        verified: "false"
      }
    ]

    userData = {
      jwt: "jwt"
    }

  })

  afterEach(() => {
    clearUserList.mockClear()
  })

  it("should call setUserList on mount", () => {
    const component = render(
      <Provider store={store}>
        <Router history={history}>
          <UserList
            listOfUsers={listOfUsers}
            userData={userData}
            setUserList={setUserList}
            clearUserList={clearUserList}
            setNotification={setNotification}
          />
        </Router>
      </Provider>
    )

    expect(setUserList).toHaveBeenCalledTimes(1)
  })

  it("should render a list of users", () => {
    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <UserList
            listOfUsers={listOfUsers}
            userData={userData}
            setUserList={setUserList}
            clearUserList={clearUserList}
            setNotification={setNotification}
          />
        </Router>
      </Provider>
    )

    expect(component).toMatchSnapshot()
  })

  it("should call clearUserList on unmount", () => {
    const component = renderer.create(
      <Provider store={store}>
        <Router history={history}>
          <UserList
            listOfUsers={listOfUsers}
            userData={userData}
            setUserList={setUserList}
            clearUserList={clearUserList}
            setNotification={setNotification}
          />
        </Router>
      </Provider>
    )

    component.unmount()
    expect(clearUserList).toHaveBeenCalledTimes(1)
  })
})
