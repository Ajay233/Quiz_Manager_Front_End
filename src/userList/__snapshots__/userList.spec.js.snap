// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserList should render a list of users 1`] = `
<div
  className="componentContainer"
>
  <div
    className="title-large"
  >
    List of all users
  </div>
  <div
    className="userDetailHeaders userHeadings"
  >
    <div
      className="forenameHeader"
    >
      First Name
    </div>
    <div
      className="surnameHeader"
    >
      Surname
    </div>
    <div
      className="emailHeader"
    >
      Email
    </div>
    <div
      className="permissionHeader"
    >
      Permission
    </div>
    <div
      className="verifiedHeader"
    >
      Verified?
    </div>
    <div
      className="deleteUserHeader"
    />
  </div>
  <div
    className="userDetails userContainer"
  >
    <div
      className="forename"
    >
      testForename1
    </div>
    <div
      className="surname"
    >
      testSurname1
    </div>
    <div
      className="email"
    >
      testEmail1
    </div>
    <div
      className="permission"
    >
      USER
    </div>
    <div
      className="verified"
    >
      <i
        className="far fa-check-circle green"
      />
    </div>
    <div
      className="deleteUser"
      onClick={[Function]}
    >
      <i
        className="fas fa-trash-alt red"
      />
       Delete
    </div>
  </div>
  <div
    className="userDetails userContainer"
  >
    <div
      className="forename"
    >
      testForename2
    </div>
    <div
      className="surname"
    >
      testSurname2
    </div>
    <div
      className="email"
    >
      testEmail2
    </div>
    <div
      className="permission"
    >
      USER
    </div>
    <div
      className="verified"
    >
      <i
        className="far fa-check-circle green"
      />
    </div>
    <div
      className="deleteUser"
      onClick={[Function]}
    >
      <i
        className="fas fa-trash-alt red"
      />
       Delete
    </div>
  </div>
</div>
`;
